bindings/wasm/src/credential/credential_builder.rs

```diff
@@ -0,0 +1,161 @@
+use identity::credential::RefreshService;
+use identity::credential::Schema;
+use identity::credential::Status;
+use identity::credential::Subject;
+use wasm_bindgen::prelude::*;
+
+use proc_typescript::typescript;
+
+use crate::error::WasmResult;
+
+impl TryFrom<ICredential> for CredentialBuilder {
+  type Error = JsValue;
+
+  fn try_from(values: ICredential) -> std::result::Result<Self, Self::Error> {
+    let ICredentialHelper {
+      context,
+      id,
+      r#type,
+      credential_subject,
+      issuer,
+      issuance_date,
+      expiration_date,
+      credential_status,
+      credential_schema,
+      refresh_service,
+      terms_of_use,
+      evidence,
+      non_transferable,
+      properties,
+    } = values.into_serde::<ICredentialHelper>().wasm_result()?;
+
+    let mut builder: CredentialBuilder = CredentialBuilder::new(properties);
+
+    if let Some(context) = context {
+      for value in context.into_vec() {
+        builder = builder.context(value);
+      }
+    }
+    if let Some(id) = id {
+      builder = builder.id(Url::parse(&id).wasm_result()?);
+    }
+    if let Some(types) = r#type {
+      for value in types.iter() {
+        builder = builder.type_(value);
+      }
+    }
+    if let Some(credential_subject) = credential_subject {
+      for subject in credential_subject.into_vec() {
+        builder = builder.subject(subject);
+      }
+    }
+    if let Some(issuer) = issuer {
+      builder = builder.issuer(issuer);
+    }
+    if let Some(issuance_date) = issuance_date {
+      builder = builder.issuance_date(issuance_date);
+    }
+    if let Some(expiration_date) = expiration_date {
+      builder = builder.expiration_date(expiration_date);
+    }
+    if let Some(credential_status) = credential_status {
+      for status in credential_status.into_vec() {
+        builder = builder.status(status);
+      }
+    }
+    if let Some(credential_schema) = credential_schema {
+      for schema in credential_schema.into_vec() {
+        builder = builder.schema(schema);
+      }
+    }
+    if let Some(refresh_service) = refresh_service {
+      for service in refresh_service.into_vec() {
+        builder = builder.refresh_service(service);
+      }
+    }
+    if let Some(terms_of_use) = terms_of_use {
+      for policy in terms_of_use.into_vec() {
+        builder = builder.terms_of_use(policy);
+      }
+    }
+    if let Some(evidence) = evidence {
+      for value in evidence.into_vec() {
+        builder = builder.evidence(value);
+      }
+    }
+    if let Some(non_transferable) = non_transferable {
+      builder = builder.non_transferable(non_transferable);
+    }
+
+    Ok(builder)
+  }
+}
+
+#[wasm_bindgen]
+extern "C" {
+  #[wasm_bindgen(typescript_type = "ICredential")]
+  pub type ICredential;
+}
+
+/// Fields for constructing a new {@link Credential}.
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[typescript(name = "ICredential", readonly, optional)]
```

Comment:
I would keep all the properties mutable, not sure if there's a reason to make them readonly?

Reply:
Several of the TypeScript interfaces are used as return values, where it's important to note that the fields are immutable since any changes to them wouldn't reflect on the struct in Wasm memory.

In other places such as `ICredential` it's used to somewhat convey that these are only meant to be used as function parameters, not representing the `Credential` itself (missing fields like `proof` and in a real `Credential` more fields are required than here). Basically it's to discourage other, incorrect uses of the interface. If we had basic function parameters rather than an interface they wouldn't be mutable either (because they wouldn't be a separate object), so I don't think it's a major disadvantage to keep these `readonly`.